tests_integration/nativeapp/test_metrics.py

```diff
@@ -0,0 +1,544 @@
+# limitations under the License.
+from shlex import split
+from typing import Dict, Callable
+from unittest import mock
+
+from snowflake.cli._app.telemetry import TelemetryEvent, CLITelemetryField
+from snowflake.cli.api.metrics import CLICounterField, CLIMetricsSpan
+from tests.nativeapp.factories import (
+    ProjectV11Factory,
+    ProjectV2Factory,
+    ApplicationPackageEntityModelFactory,
+    ApplicationEntityModelFactory,
+)
+from tests.project.fixtures import *
+from tests_common import temp_dir
+from tests_integration.test_utils import extract_first_telemetry_message_of_type
+
+
+@pytest.mark.integration
+@pytest.mark.parametrize(
+    "command,expected_counter",
+    [
+        (
+            [
+                "sql",
+                "-q",
+                "select '<% ctx.env.test %>'",
+                "--env",
+                "test=value_from_cli",
+            ],
+            1,
+        ),
+        (["sql", "-q", "select 'string'"], 0),
+    ],
+)
+@mock.patch("snowflake.connector.telemetry.TelemetryClient.try_add_log_to_batch")
+def test_sql_templating_emits_counter(
+    mock_telemetry,
+    command: List[str],
+    expected_counter,
+    runner,
+):
+    result = runner.invoke_with_connection_json(command)
+
+    assert result.exit_code == 0
+
+    message = extract_first_telemetry_message_of_type(
+        mock_telemetry, TelemetryEvent.CMD_EXECUTION_RESULT.value
+    )
+
+    assert (
+        message[CLITelemetryField.COUNTERS.value][CLICounterField.SQL_TEMPLATES]
+        == expected_counter
+    )
+
+
+@pytest.mark.integration
+@mock.patch("snowflake.connector.telemetry.TelemetryClient.try_add_log_to_batch")
+def test_feature_counters_v1_post_deploy_set_and_package_scripts_available(
+    mock_telemetry,
+    runner,
+    nativeapp_teardown,
+    temp_dir,
+):
+    ProjectV11Factory(
+        pdf__native_app__artifacts=["README.md", "setup.sql", "manifest.yml"],
+        pdf__native_app__package__post_deploy=[
+            {"sql_script": "post_deploy1.sql"},
+        ],
+        files={
+            "README.md": "",
+            "setup.sql": "select 1",
+            "manifest.yml": "\n",
+            "post_deploy1.sql": "\n",
+        },
+    )
+
+    with nativeapp_teardown(project_dir=Path(temp_dir)):
+        runner.invoke_with_connection(["app", "deploy"])
+
+        message = extract_first_telemetry_message_of_type(
+            mock_telemetry, TelemetryEvent.CMD_EXECUTION_RESULT.value
+        )
+
+        assert message[CLITelemetryField.COUNTERS.value] == {
+            CLICounterField.SNOWPARK_PROCESSOR: 0,
+            CLICounterField.TEMPLATES_PROCESSOR: 0,
+            CLICounterField.PDF_TEMPLATES: 0,
+            CLICounterField.POST_DEPLOY_SCRIPTS: 1,
+            CLICounterField.PACKAGE_SCRIPTS: 0,
+        }
+
+
+@pytest.mark.integration
+@mock.patch("snowflake.connector.telemetry.TelemetryClient.try_add_log_to_batch")
+def test_feature_counters_v2_post_deploy_not_available_in_bundle(
```

Comment:
counter tests moved from `test_feature_metrics.py` to consolidate them in one file and keep to the convention, and converted to use factories